File: examples/python/m2-tool-runtime/detect.py
"""单步 postcondition checker：检查一次 ToolResult 与 workspace 状态。"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any

from tool_types import ToolResult
from workspace import Workspace


class ExpectationKind(str, Enum):
    TOOL_SUCCEEDED = "tool_succeeded"
    TOOL_FAILED = "tool_failed"
    OUTPUT_CONTAINS = "output_contains"
    FILE_CONTAINS = "file_contains"
    FILE_NOT_CONTAINS = "file_not_contains"
    FILE_EXISTS = "file_exists"


@dataclass(frozen=True)
class Expectation:
    kind: ExpectationKind
    text: str | None = None
    path: str | None = None
    call_id: str | None = None
    error_substring: str | None = None


@dataclass(frozen=True)
class CheckResult:
    expectation: Expectation
    passed: bool
    detail: str


@dataclass(frozen=True)
class DetectionReport:
    checks: tuple[CheckResult, ...]

    @property
    def passed(self) -> bool:
        return all(check.passed for check in self.checks)

    def summary(self) -> str:
        status = "passed" if self.passed else "failed"
        lines = [f"detection={status} checks={len(self.checks)}"]
        for check in self.checks:
            mark = "ok" if check.passed else "FAIL"
            lines.append(f"  [{mark}] {check.expectation.kind.value}: {check.detail}")
        return "\n".join(lines)


def detect(
    workspace: Workspace,
    result: ToolResult,
    expectations: list[Expectation],
) -> DetectionReport:
    """对单次工具结果运行局部后置条件；不生成 M7 任务级 Evidence。"""

    checks = [
        _evaluate(workspace, result, expectation) for expectation in expectations
    ]
    return DetectionReport(checks=tuple(checks))


def _evaluate(
    workspace: Workspace,
    result: ToolResult,
    expectation: Expectation,
) -> CheckResult:
    kind = expectation.kind

    if kind is ExpectationKind.TOOL_SUCCEEDED:
        if expectation.call_id and result.call_id != expectation.call_id:
            return CheckResult(expectation, False, "call_id mismatch")
        ok = result.succeeded
        return CheckResult(
            expectation,
            ok,
            "tool succeeded" if ok else f"tool failed: {result.error}",
        )

    if kind is ExpectationKind.TOOL_FAILED:
        if expectation.call_id and result.call_id != expectation.call_id:
            return CheckResult(expectation, False, "call_id mismatch")
        if result.succeeded:
            return CheckResult(expectation, False, "expected failure, got success")
        if expectation.error_substring:
            haystack = result.error or ""
            ok = expectation.error_substring in haystack
            return CheckResult(
                expectation,
                ok,
                "error matched" if ok else f"error did not contain {expectation.error_substring!r}",
            )
        return CheckResult(expectation, True, f"failed as expected: {result.error}")

    if kind is ExpectationKind.OUTPUT_CONTAINS:
        text = expectation.text or ""
        blob = _output_blob(result)
        ok = text in blob
        return CheckResult(
            expectation,
            ok,
            "output contains text" if ok else f"output missing {text!r}",
        )

    if kind is ExpectationKind.FILE_EXISTS:
        path = expectation.path or ""
        try:
            target = workspace.resolve(path, must_exist=False)
            ok = target.exists()
        except Exception as error:  # noqa: BLE001
            return CheckResult(expectation, False, str(error))
        return CheckResult(
            expectation,
            ok,
            "file exists" if ok else f"missing file {path!r}",
        )

    if kind is ExpectationKind.FILE_CONTAINS:
        return _file_text_check(workspace, expectation, should_contain=True)

    if kind is ExpectationKind.FILE_NOT_CONTAINS:
        return _file_text_check(workspace, expectation, should_contain=False)

    return CheckResult(expectation, False, f"unknown expectation: {kind}")


def _file_text_check(
    workspace: Workspace,
    expectation: Expectation,
    *,
    should_contain: bool,
) -> CheckResult:
    path = expectation.path or ""
    text = expectation.text or ""
    try:
        target = workspace.resolve(path, must_exist=True)
        content = target.read_text(encoding="utf-8")
    except Exception as error:  # noqa: BLE001
        return CheckResult(expectation, False, str(error))

    contains = text in content
    ok = contains if should_contain else not contains
    if should_contain:
        detail = "file contains text" if ok else f"file missing {text!r}"
    else:
        detail = "file does not contain text" if ok else f"file still contains {text!r}"
    return CheckResult(expectation, ok, detail)


def _output_blob(result: ToolResult) -> str:
    if not result.succeeded:
        return result.error or ""
    output: Any = result.output
    if output is None:
        return ""
    if isinstance(output, str):
        return output
    import json

    return json.dumps(output, ensure_ascii=False)

File: examples/rust/m2-tool-runtime/src/detect.rs
//! 单步 postcondition checker；不等同于 M7 任务级 Validation/Evidence。

use crate::tool_types::ToolResult;
use crate::workspace::Workspace;
use std::fmt;

/// 单步可检查的后置条件种类。
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum ExpectationKind {
    ToolSucceeded,
    ToolFailed,
    OutputContains,
    FileContains,
    FileNotContains,
    FileExists,
}

/// 一条局部后置条件。
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Expectation {
    pub kind: ExpectationKind,
    pub text: Option<String>,
    pub path: Option<String>,
    pub call_id: Option<String>,
    pub error_substring: Option<String>,
}

/// 单项检查结果。
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CheckResult {
    pub expectation: Expectation,
    pub passed: bool,
    pub detail: String,
}

/// 一次工具调用后的全部检查结果。
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DetectionReport {
    pub checks: Vec<CheckResult>,
}

impl DetectionReport {
    /// 是否所有局部检查都通过。
    pub fn passed(&self) -> bool {
        self.checks.iter().all(|check| check.passed)
    }

    /// 生成面向读者的多行摘要。
    pub fn summary(&self) -> String {
        let status = if self.passed() { "passed" } else { "failed" };
        let mut lines = vec![format!("detection={status} checks={}", self.checks.len())];
        for check in &self.checks {
            let mark = if check.passed { "ok" } else { "FAIL" };
            lines.push(format!(
                "  [{mark}] {}: {}",
                kind_name(&check.expectation.kind),
                check.detail
            ));
        }
        lines.join("\n")
    }
}

impl fmt::Display for DetectionReport {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.summary())
    }
}

/// 对一次结果和当前 Workspace 运行局部后置条件。
pub fn detect(
    workspace: &Workspace,
    result: &ToolResult,
    expectations: &[Expectation],
) -> DetectionReport {
    let checks = expectations
        .iter()
        .map(|expectation| evaluate(workspace, result, expectation))
        .collect();
    DetectionReport { checks }
}

fn evaluate(workspace: &Workspace, result: &ToolResult, expectation: &Expectation) -> CheckResult {
    match expectation.kind {
        ExpectationKind::ToolSucceeded => {
            if expectation
                .call_id
                .as_ref()
                .is_some_and(|call_id| call_id != &result.call_id)
            {
                return CheckResult {
                    expectation: expectation.clone(),
                    passed: false,
                    detail: "call_id mismatch".into(),
                };
            }
            let ok = result.succeeded();
            CheckResult {
                expectation: expectation.clone(),
                passed: ok,
                detail: if ok {
                    "tool succeeded".into()
                } else {
                    format!("tool failed: {:?}", result.error)
                },
            }
        }
        ExpectationKind::ToolFailed => {
            if expectation
                .call_id
                .as_ref()
                .is_some_and(|call_id| call_id != &result.call_id)
            {
                return CheckResult {
                    expectation: expectation.clone(),
                    passed: false,
                    detail: "call_id mismatch".into(),
                };
            }
            if result.succeeded() {
                return CheckResult {
                    expectation: expectation.clone(),
                    passed: false,
                    detail: "expected failure, got success".into(),
                };
            }
            if let Some(sub) = &expectation.error_substring {
                let haystack = result.error.clone().unwrap_or_default();
                let ok = haystack.contains(sub);
                return CheckResult {
                    expectation: expectation.clone(),
                    passed: ok,
                    detail: if ok {
                        "error matched".into()
                    } else {
                        format!("error did not contain {sub:?}")
                    },
                };
            }
            CheckResult {
                expectation: expectation.clone(),
                passed: true,
                detail: format!("failed as expected: {:?}", result.error),
            }
        }
        ExpectationKind::OutputContains => {
            let text = expectation.text.clone().unwrap_or_default();
            let blob = output_blob(result);
            let ok = blob.contains(&text);
            CheckResult {
                expectation: expectation.clone(),
                passed: ok,
                detail: if ok {
                    "output contains text".into()
                } else {
                    format!("output missing {text:?}")
                },
            }
        }
        ExpectationKind::FileExists => {
            let path = expectation.path.clone().unwrap_or_default();
            match workspace.resolve(Some(&path), false) {
                Ok(target) => {
                    let ok = target.exists();
                    CheckResult {
                        expectation: expectation.clone(),
                        passed: ok,
                        detail: if ok {
                            "file exists".into()
                        } else {
                            format!("missing file {path:?}")
                        },
                    }
                }
                Err(error) => CheckResult {
                    expectation: expectation.clone(),
                    passed: false,
                    detail: error.to_string(),
                },
            }
        }
        ExpectationKind::FileContains | ExpectationKind::FileNotContains => file_text_check(
            workspace,
            expectation,
            expectation.kind == ExpectationKind::FileContains,
        ),
    }
}

fn file_text_check(
    workspace: &Workspace,
    expectation: &Expectation,
    should_contain: bool,
) -> CheckResult {
    let path = expectation.path.clone().unwrap_or_default();
    let text = expectation.text.clone().unwrap_or_default();
    match workspace.resolve(Some(&path), true) {
        Ok(target) => match std::fs::read_to_string(target) {
            Ok(content) => {
                let contains = content.contains(&text);
                let ok = if should_contain { contains } else { !contains };
                let detail = if should_contain {
                    if ok {
                        "file contains text".into()
                    } else {
                        format!("file missing {text:?}")
                    }
                } else if ok {
                    "file does not contain text".into()
                } else {
                    format!("file still contains {text:?}")
                };
                CheckResult {
                    expectation: expectation.clone(),
                    passed: ok,
                    detail,
                }
            }
            Err(error) => CheckResult {
                expectation: expectation.clone(),
                passed: false,
                detail: error.to_string(),
            },
        },
        Err(error) => CheckResult {
            expectation: expectation.clone(),
            passed: false,
            detail: error.to_string(),
        },
    }
}

fn output_blob(result: &ToolResult) -> String {
    if !result.succeeded() {
        return result.error.clone().unwrap_or_default();
    }
    match &result.output {
        None => String::new(),
        Some(serde_json::Value::String(text)) => text.clone(),
        Some(value) => value.to_string(),
    }
}

fn kind_name(kind: &ExpectationKind) -> &'static str {
    match kind {
        ExpectationKind::ToolSucceeded => "tool_succeeded",
        ExpectationKind::ToolFailed => "tool_failed",
        ExpectationKind::OutputContains => "output_contains",
        ExpectationKind::FileContains => "file_contains",
        ExpectationKind::FileNotContains => "file_not_contains",
        ExpectationKind::FileExists => "file_exists",
    }
}
